File: 8.字符串转换整数-atoi.py
#
# @lc app=leetcode.cn id=8 lang=python3
#
# [8] 字符串转换整数 (atoi)
#

# @lc code=start

import re


class Solution:
    def myAtoi(self, str: str) -> int:
        # 我的
        # MAX = 2147483647
        # MIN = -2147483648
        # flags = {'+', '-'}
        # read_flag = True
        # stack = []
        # for ch in str:
        #     if read_flag:
        #         if ch == ' ':
        #             continue
        #         if ch in flags or ch.isdigit():
        #             stack.append(ch)
        #             read_flag = False
        #         else:
        #             return 0
        #     else:
        #         if ch.isdigit():
        #             stack.append(ch)
        #         else:
        #             break
        # wei = 0
        # res = 0
        # while stack:
        #     ch = stack.pop()
        #     if ch.isdigit():
        #         res += int(ch) * 10 ** wei
        #         wei += 1
        #     elif ch == '-':
        #         res = -res
        # res = MAX if res > MAX else res
        # res = MIN if res < MIN else res

        # return res
        MAX = 2147483647
        MIN = -2147483648
        str = str.lstrip()
        num_re = re.compile(r'^[\+\-]?\d+')
        res = num_re.findall(str)  # 查找匹配的内容
        res = int(*res)  # 由于返回的是个列表，解包并且转换成整数
        return max(min(res, MAX), MIN)  # 返回值)

# @lc code=end

File: 8.字符串转换整数-atoi.java
/*
 * @lc app=leetcode.cn id=8 lang=java
 *
 * [8] 字符串转换整数 (atoi)
 */

// @lc code=start
import java.util.Stack;
class Solution {
    public static int myAtoi(String str) {
        long MAX = 2147483647;
        long MIN = -2147483648;
        boolean readFlag = true;
        Stack<Character> stack = new Stack<>();
        for (char ch : str.toCharArray()) {
            if (readFlag) {
                if (ch == ' ')
                    continue;
                else if (ch == '-' || ch == '+' || Character.isDigit(ch)) {
                    stack.push(ch);
                    readFlag = false;
                } else
                    return 0;
            } else {
                if (Character.isDigit(ch))
                    stack.push(ch);
                else
                    break;
            }
        }
        long res = 0;
        int wei = 0;
        char ch;
        double tmp;
        while (!stack.empty()) {
            ch = stack.pop();
            if (Character.isDigit(ch)) {
                tmp = Integer.parseInt(String.valueOf(ch)) * Math.pow(10, wei);
                if (tmp + res > MAX) {
                    res = MAX;
                }
                res += tmp;
                wei++;
            } else if (ch == '-')
                res = -res;
        }
        return (int) Math.max(Math.min(res, MAX), MIN);
    }
}
// @lc code=end

